### PR TITLE
Apply PHP 7.4 syntax and typed property

### PR DESCRIPTION
**Pull request by samsonasik**
Signed-off-by: Abdul Malik Ikhsan <samsonasik@gmail.com>

|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

Since composer.json require php 7.4, php 7.4 syntax can be applied with typed properties. 

- For `final class`, typed properties for protected is allowed if no inherit of parent or used by its final class parent.
- Otherwise, update only private properties.

### CODE REVIEW

File: src/Filter/FilterComposite.php [hunk 1]
```diff
@@ -25,11 +25,9 @@ final class FilterComposite implements FilterInterface
      */
     public const CONDITION_AND = 2;
 
-    /** @var ArrayObject */
-    protected $andFilter;
+    protected ArrayObject $andFilter;
```
Comment (Ocramius) [inline src/Filter/FilterComposite.php]: Careful with `protected` stuff :|

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Filter/FilterComposite.php (hunk 1 above)
Comment (samsonasik) [reply src/Filter/FilterComposite.php]: it is final class, and doesn't has parent property usage.

Comment (Ocramius) [reply src/Filter/FilterComposite.php]: Oooh, didn't see - should we change these to `private`, when we encounter them? :thinking:

Comment (samsonasik) [reply src/Filter/FilterComposite.php]: yes, updated to private https://github.com/laminas/laminas-hydrator/pull/91/commits/93f4af2165b1b2788372beedd6bb5f95ed8371ed

**Comment by Ocramius** [comment]
I think the psalm issues should probably go to baseline :thinking:

**Comment by samsonasik** [comment]
@Ocramius updated to psalm-baseline https://github.com/laminas/laminas-hydrator/pull/91/commits/092c563350392d0886d91ad390834f88995bc872

**Comment by samsonasik** [comment]
@Ocramius all green 🎉

File: src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php [hunk 2]
```diff
@@ -71,14 +71,8 @@ private function getPatternAndReplacement(): array
...
-            }
-            : function ($value) {
-                // ignore unicode characters w/ strtolower
-                return preg_replace_callback('#([A-Z])#', function ($matches) {
-                    return strtolower($matches[1]);
-                }, $value);
-            };
+            ? fn($value) => mb_strtolower($value, 'UTF-8')
+            : fn($value) => // ignore unicode characters w/ strtolower
+preg_replace_callback('#([A-Z])#', fn($matches) => strtolower($matches[1]), $value);
```
Comment (Ocramius) [inline src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php]: Must say that some of these alignment changes are really strange :|

File: src/NamingStrategy/CompositeNamingStrategy.php [hunk 3]
```diff
@@ -4,26 +4,25 @@
...
      */
     public function __construct(array $strategies, ?NamingStrategyInterface $defaultNamingStrategy = null)
     {
         $this->namingStrategies = array_map(
-            function (NamingStrategyInterface $strategy) {
-                // this callback is here only to ensure type-safety
-                return $strategy;
-            },
+            fn(NamingStrategyInterface $strategy) => // this callback is here only to ensure type-safety
+            $strategy,
```
Comment (Ocramius) [inline src/NamingStrategy/CompositeNamingStrategy.php]: I would expect `$strategy` to be on the same line - didn't even notice that it was all one closure :scream:

File: src/NamingStrategy/CompositeNamingStrategy.php [hunk 4]
```diff
@@ -4,26 +4,25 @@
...
      * @param NamingStrategyInterface[]    $strategies            indexed by the name they translate
      */
     public function __construct(array $strategies, ?NamingStrategyInterface $defaultNamingStrategy = null)
     {
         $this->namingStrategies = array_map(
-            function (NamingStrategyInterface $strategy) {
-                // this callback is here only to ensure type-safety
-                return $strategy;
-            },
+            fn(NamingStrategyInterface $strategy) => // this callback is here only to ensure type-safety
```
Comment (Ocramius) [inline src/NamingStrategy/CompositeNamingStrategy.php]: We should use `static fn` where possible

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php [hunk 5]
```diff
@@ -54,9 +54,7 @@ private function getPatternAndReplacement(string $pregQuotedSeparator): PcreRepl
             ? $this->getUnicodePatternAndReplacement($pregQuotedSeparator)
             : new PcreReplacement(
                 '#(' . $pregQuotedSeparator . ')([\S]{1})#',
-                function ($matches) {
-                    return strtoupper($matches[2]);
-                }
+                fn($matches) => strtoupper($matches[2])
```
Comment (Ocramius) [inline src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: `static`

Comment (Ocramius) [reply src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: We can add `: string`

File: src/Filter/OptionalParametersFilter.php [hunk 6]
```diff
@@ -53,9 +53,7 @@ public function filter(string $property, ?object $instance = null): bool
 
         $mandatoryParameters = array_filter(
             $reflectionMethod->getParameters(),
-            function (ReflectionParameter $parameter) {
-                return ! $parameter->isOptional();
-            }
+            fn(ReflectionParameter $parameter) => ! $parameter->isOptional()
```
Comment (Ocramius) [inline src/Filter/OptionalParametersFilter.php]: `: bool` and `static`

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php [hunk 7]
```diff
@@ -65,25 +63,19 @@ private function getUnicodePatternAndReplacement(string $pregQuotedSeparator): P
         return $this->hasMbStringSupport()
             ? new PcreReplacement(
                 '#(' . $pregQuotedSeparator . ')(\P{Z}{1})#u',
-                function ($matches) {
-                    return mb_strtoupper($matches[2], 'UTF-8');
-                }
+                fn($matches) => mb_strtoupper($matches[2], 'UTF-8')
```
Comment (Ocramius) [inline src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: `: string` and `static`

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php [hunk 8]
```diff
@@ -65,25 +63,19 @@ private function getUnicodePatternAndReplacement(string $pregQuotedSeparator): P
...
             )
             : new PcreReplacement(
                 '#(' . $pregQuotedSeparator . ')'
                     . '([^\p{Z}\p{Ll}]{1}|[a-zA-Z]{1})#u',
-                function ($matches) {
-                    return strtoupper($matches[2]);
-                }
+                fn($matches) => strtoupper($matches[2])
```
Comment (Ocramius) [inline src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: `: string` and `static`

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php [hunk 9]
```diff
@@ -65,25 +63,19 @@ private function getUnicodePatternAndReplacement(string $pregQuotedSeparator): P
...
             );
     }
 
     private function getLcFirstFunction(): callable
     {
         return $this->hasMbStringSupport()
-            ? function ($value) {
-                return mb_strtolower($value[0], 'UTF-8') . substr($value, 1);
-            }
+            ? fn($value) => mb_strtolower($value[0], 'UTF-8') . substr($value, 1)
```
Comment (Ocramius) [inline src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: `: string` and `static`

File: src/ObjectPropertyHydrator.php [hunk 10]
```diff
@@ -63,9 +63,7 @@ public function hydrate(array $data, object $object)
             $reflection = new ReflectionClass($object);
             $properties = array_fill_keys(
                 array_map(
-                    function (ReflectionProperty $property) {
-                        return $property->getName();
-                    },
+                    fn(ReflectionProperty $property) => $property->getName(),
```
Comment (Ocramius) [inline src/ObjectPropertyHydrator.php]: `: string` and `static`

File: src/StandaloneHydratorPluginManager.php [hunk 11]
```diff
@@ -62,14 +62,12 @@ final class StandaloneHydratorPluginManager implements HydratorPluginManagerInte
...
-    private $factories = [];
+    private array $factories = [];
 
     public function __construct()
     {
         /** @psalm-suppress UnusedClosureParam */
-        $invokableFactory = function (ContainerInterface $container, string $class): object {
-            return new $class();
-        };
+        $invokableFactory = fn(ContainerInterface $container, string $class): object => new $class();
```
Comment (Ocramius) [inline src/StandaloneHydratorPluginManager.php]: `static`

File: src/Strategy/CollectionStrategy.php [hunk 12]
```diff
@@ -87,11 +85,9 @@ public function hydrate($value, ?array $data = null)
 
         $reflection = new ReflectionClass($this->objectClassName);
 
-        return array_map(function ($data) use ($reflection) {
-            return $this->objectHydrator->hydrate(
-                $data,
-                $reflection->newInstanceWithoutConstructor()
-            );
-        }, $value);
+        return array_map(fn($data) => $this->objectHydrator->hydrate(
```
Comment (Ocramius) [inline src/Strategy/CollectionStrategy.php]: `: object`?

File: test/HydratorClosureStrategyTest.php [hunk 13]
```diff
@@ -68,15 +66,11 @@ public function testRetrieveStrategy(): void
     public function testExtractingObjects(): void
     {
         $this->hydrator->addStrategy('field1', new ClosureStrategy(
-            function ($value) {
-                return sprintf('%s', $value);
-            },
+            fn($value) => sprintf('%s', $value),
```
Comment (Ocramius) [inline test/HydratorClosureStrategyTest.php]: `static` and `: string`

File: test/HydratorClosureStrategyTest.php [hunk 14]
```diff
@@ -68,15 +66,11 @@ public function testRetrieveStrategy(): void
...
             null
         ));
         $this->hydrator->addStrategy('field2', new ClosureStrategy(
-            function ($value) {
-                return sprintf('hello, %s!', $value);
-            },
+            fn($value) => sprintf('hello, %s!', $value),
```
Comment (Ocramius) [inline test/HydratorClosureStrategyTest.php]: `static` and `: string`

File: test/HydratorClosureStrategyTest.php [hunk 15]
```diff
@@ -91,15 +85,11 @@ public function testHydratingObjects(): void
     {
         $this->hydrator->addStrategy('field2', new ClosureStrategy(
             null,
-            function ($value) {
-                return sprintf('hello, %s!', $value);
-            }
+            fn($value) => sprintf('hello, %s!', $value)
```
Comment (Ocramius) [inline test/HydratorClosureStrategyTest.php]: `static` and `: string`

File: test/HydratorClosureStrategyTest.php [hunk 16]
```diff
@@ -91,15 +85,11 @@ public function testHydratingObjects(): void
...
         ));
         $this->hydrator->addStrategy('field3', new ClosureStrategy(
             null,
-            function ($value) {
-                return new TestAsset\HydratorClosureStrategyEntity($value, sprintf('111%s', $value));
-            }
+            fn($value) => new TestAsset\HydratorClosureStrategyEntity($value, sprintf('111%s', $value))
```
Comment (Ocramius) [inline test/HydratorClosureStrategyTest.php]: `static` and `: object`?

File: test/HydratorPluginManagerFactoryTest.php [hunk 17]
```diff
@@ -50,9 +50,7 @@ public function testConfiguresHydratorServicesWhenFound(): void
                 ],
                 'factories' => [
                     /** @psalm-return MockObject&HydratorInterface */
-                    'test-too' => function () use ($hydrator): HydratorInterface {
-                        return $hydrator;
-                    },
+                    'test-too' => fn(): HydratorInterface => $hydrator,
```
Comment (Ocramius) [inline test/HydratorPluginManagerFactoryTest.php]: `static`

File: test/Strategy/CollectionStrategyTest.php [hunk 18]
```diff
@@ -179,11 +179,9 @@ public function testExtractUsesHydratorToExtractValues(): void
         $extraction =  * @return string[]
         $extraction =  * @psalm-return array{value: string}
          */
-        function (TestAsset\User $value): array {
-            return [
-                'value' => spl_object_hash($value),
-            ];
-        };
+        fn(TestAsset\User $value): array => [
```
Comment (Ocramius) [inline test/Strategy/CollectionStrategyTest.php]: `static`

File: test/Strategy/StrategyChainTest.php [hunk 19]
```diff
@@ -24,42 +24,30 @@ public function testExtract(): void
     {
         $chain = new StrategyChain([
             new ClosureStrategy(
-                function ($value) {
-                    return $value % 12;
-                }
+                fn($value) => $value % 12
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 20]
```diff
@@ -24,42 +24,30 @@ public function testExtract(): void
...
             ),
             new ClosureStrategy(
-                function ($value) {
-                    return $value % 9;
-                }
+                fn($value) => $value % 9
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 21]
```diff
@@ -24,42 +24,30 @@ public function testExtract(): void
...
             ),
         ]);
         $this->assertEquals(3, $chain->extract(87));
 
         $chain = new StrategyChain([
             new ClosureStrategy(
-                function ($value) {
-                    return $value % 8;
-                }
+                fn($value) => $value % 8
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 22]
```diff
@@ -24,42 +24,30 @@ public function testExtract(): void
...
             ),
             new ClosureStrategy(
-                function ($value) {
-                    return $value % 3;
-                }
+                fn($value) => $value % 3
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 23]
```diff
@@ -24,42 +24,30 @@ public function testExtract(): void
...
             ),
         ]);
         $this->assertEquals(1, $chain->extract(20));
 
         $chain = new StrategyChain([
             new ClosureStrategy(
-                function ($value) {
-                    return $value % 7;
-                }
+                fn($value) => $value % 7
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 24]
```diff
@@ -24,42 +24,30 @@ public function testExtract(): void
...
             ),
             new ClosureStrategy(
-                function ($value) {
-                    return $value % 6;
-                }
+                fn($value) => $value % 6
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 25]
```diff
@@ -70,47 +58,35 @@ public function testHydrate(): void
         $chain = new StrategyChain([
             new ClosureStrategy(
                 null,
-                function ($value) {
-                    return $value % 3;
-                }
+                fn($value) => $value % 3
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 26]
```diff
@@ -70,47 +58,35 @@ public function testHydrate(): void
...
             ),
             new ClosureStrategy(
                 null,
-                function ($value) {
-                    return $value % 7;
-                }
+                fn($value) => $value % 7
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 27]
```diff
@@ -70,47 +58,35 @@ public function testHydrate(): void
...
         ]);
         $this->assertEquals(0, $chain->hydrate(87));
 
         $chain = new StrategyChain([
             new ClosureStrategy(
                 null,
-                function ($value) {
-                    return $value % 8;
-                }
+                fn($value) => $value % 8
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 28]
```diff
@@ -70,47 +58,35 @@ public function testHydrate(): void
...
             ),
             new ClosureStrategy(
                 null,
-                function ($value) {
-                    return $value % 3;
-                }
+                fn($value) => $value % 3
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 29]
```diff
@@ -70,47 +58,35 @@ public function testHydrate(): void
...
         ]);
         $this->assertEquals(2, $chain->hydrate(20));
 
         $chain = new StrategyChain([
             new ClosureStrategy(
                 null,
-                function ($value) {
-                    return $value % 4;
-                }
+                fn($value) => $value % 4
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

File: test/Strategy/StrategyChainTest.php [hunk 30]
```diff
@@ -70,47 +58,35 @@ public function testHydrate(): void
...
             ),
             new ClosureStrategy(
                 null,
-                function ($value) {
-                    return $value % 9;
-                }
+                fn($value) => $value % 9
```
Comment (Ocramius) [inline test/Strategy/StrategyChainTest.php]: `: int` and `static`

**Review by Ocramius — CHANGES_REQUESTED** [review]
Let's fix those closures, then re-generate the baseline again :D 

Sorry for the nitpicking, it's just that, since this is a CS change, we may as well do the CS change well :+1:

**Comment by samsonasik** [comment]
@Ocramius updated for static ArrowFunction and apply return type declaration on ArrowFunction.

**Comment by samsonasik** [comment]
@Ocramius all green 🎉

**Review by Ocramius — APPROVED** [review]
Thanks @samsonasik!

**Comment by Ocramius** [comment]
Ah, I see that 6317c17 was pushed after I merged, sorry :|

**Pull request merged**